virttest/qemu_storage.py

```diff
@@ -1096,6 +1096,132 @@ def snapshot_apply(self):
+            if granularity:
+                cmd += ' -g %s' % granularity
+        else:
+            raise exceptions.TestError(
+                "Image format %s is not qcow2, check please." %
+                self.image_format)
+
+        process.run(cmd)
+
+    def bitmap_remove(self, bitmap_name):
+        """
+        Remove a bitmap from a image.
+
+        :note: params should contain:
+               bitmap_name -- the name of bitmap
+        """
+
+        cmd = self.image_cmd
+        if self.image_format == 'qcow2':
+            cmd += " bitmap --remove -f %s %s %s" % (
+                self.image_format, self.image_filename, bitmap_name)
+        else:
+            raise exceptions.TestError(
+                "Image format %s is not qcow2, check please." %
+                self.image_format)
+
+        process.run(cmd)
+
+    def bitmap_clear(self, bitmap_name):
+        """
+        Clear a bitmap of a image.
+
+        :note: params should contain:
+               bitmap_name -- the name of bitmap
+        """
+
+        cmd = self.image_cmd
+        if self.image_format == 'qcow2':
+            cmd += " bitmap --clear -f %s %s %s" % (
+                self.image_format, self.image_filename, bitmap_name)
+        else:
+            raise exceptions.TestError(
+                "Image format %s is not qcow2, check please." %
+                self.image_format)
+
+        process.run(cmd)
+
+    def bitmap_enable(self, bitmap_name):
+        """
+        Enable a bitmap of a image.
+
+        :note: params should contain:
+               bitmap_name -- the name of bitmap
+        """
+
+        cmd = self.image_cmd
+        if self.image_format == 'qcow2':
+            cmd += " bitmap --enable -f %s %s %s" % (
+                self.image_format, self.image_filename, bitmap_name)
+        else:
+            raise exceptions.TestError(
+                "Image format %s is not qcow2, check please." %
+                self.image_format)
+
+        process.run(cmd)
+
+    def bitmap_disable(self, bitmap_name):
+        """
+        Disable a bitmap of a image.
+
+        :note: params should contain:
+                bitmap_name -- the name of bitmap
+        """
+
+        cmd = self.image_cmd
+        if self.image_format == 'qcow2':
+            cmd += " bitmap --disable -f %s %s %s" % (
+                self.image_format, self.image_filename, bitmap_name)
+        else:
+            raise exceptions.TestError(
+                "Image format %s is not qcow2, check please." %
+                self.image_format)
+
+        process.run(cmd)
+
+    def bitmap_merge(self, bitmap_name_source, bitmap_name, params, root_dir):
+        """
+        Merge a bitmap from A image to B image.
+
+        :note: params should contain:
+                source_bitmap -- the source bitmap in source file
+                bitmap_name -- the name of bitmap
+        """
+
+        bitmap_source = params["bitmap_source"]
```

Comment:
bitmap_params = self.params.object_params(bitmap_name_source)
bitmap_source = bitmap_params.get('bitmap_source', self.tag)

bitmap_source should be a bitmap level param, e.g. 
bitmap_source_bm1 = image1
bitmap_source_bm2 = data1

Reply:
please also update base.cfg with
bitmap_granularity: xxx
bitmap_source:  the image where the bitmap is added, e.g. image1/data1